src/components/app/index.tsx

```diff
@@ -25,33 +66,131 @@ export default class App extends Component<Props, State> {
     }
   }
 
+  getEncoderName(type: ImageType, options: CodecOptions) {
+    const name = EncoderNames[type];
+    if (typeof name === 'function') return name(options);
+    return name;
+  }
+
+  @bind
+  setLeftType(leftType: ImageType) {
+    this.setState({ leftType, leftOptions: {} });
+    this.optionsUpdated();
+  }
+
+  @bind
+  setRightType(rightType: ImageType) {
+    this.setState({ rightType, rightOptions: {} });
+    this.optionsUpdated();
+  }
+
+  @bind
+  setLeftOptions(leftOptions: any) {
+    this.setState({ leftOptions });
+    this.optionsUpdated();
+  }
+
+  @bind
+  setRightOptions(rightOptions: any) {
+    this.setState({ rightOptions });
+    this.optionsUpdated();
+  }
+
+  optionsUpdated() {
+    if (!this.optionsUpdateTimer) {
+      this.optionsUpdateTimer = setTimeout(this.commitOptionsUpdated, 500);
+    }
+  }
+
+  @bind
+  commitOptionsUpdated() {
+    this.optionsUpdateTimer = null;
+    this.updateImages();
+  }
+
   @bind
   async onFileChange(event: Event) {
+    this.setState({ loading: true });
     const fileInput = event.target as HTMLInputElement;
     if (!fileInput.files || !fileInput.files[0]) return;
     // TODO: handle decode error
-    const bitmap = await createImageBitmap(fileInput.files[0]);
-    const data = await bitmapToImageData(bitmap);
-    const encoder = new MozJpegEncoder();
-    const compressedData = await encoder.encode(data);
-    const blob = new Blob([compressedData], {type: 'image/jpeg'});
-    const compressedImage = await createImageBitmap(blob);
-    this.setState({ img: compressedImage });
+    const sourceImg = await createImageBitmap(fileInput.files[0]);
+    // compute the corresponding ImageData once since it only changes when the file changes:
+    const sourceData = await bitmapToImageData(sourceImg);
+    this.setState({ sourceImg, sourceData, loading: false });
+    this.updateImages();
   }
 
-  render({ }: Props, { img }: State) {
+  async updateImages() {
+    const { sourceData, leftType, rightType, leftOptions, rightOptions } = this.state;
+    if (!sourceData) return;
+    const id = ++this.compressCounter;
+    this.setState({ loading: true });
+    const [leftImg, rightImg] = await Promise.all([
+      this.updateCompressedImage(sourceData, leftType, leftOptions),
+      this.updateCompressedImage(sourceData, rightType, rightOptions)
+    ]);
+    if (this.compressCounter !== id) return;
+    this.setState({ leftImg, rightImg, loading: false });
+  }
+
+  async updateCompressedImage(sourceData: ImageData, type: ImageType, options: CodecOptions, retries= 0) {
+    if (type === 'original') {
```

Comment:
Just wondering: Worth creating a pseudo-encode in the `AllEncoders` dictionary that’s just a pass-through?

Reply:
Ah yeah that's cleaner.